File: 프로그래머스/lv1/42862. 체육복/체육복.js
//answer를 최솟값을 초기화 , 최솟값 n-lost.length
//obj(객체)에 체육복의 여분 정보를 저장 키가 학생의 번호, 벨류가 여분1개만
//lost를 돌아가면서 reserve에 값이 있을경우 우선시해서 여분의 갯수를 - 해줌, 그리고 answer를 ++해줌
//다시 lost를 돌아가면서 번호가 -1일경우, 번호가 +1일경우 obj에 있는지 체크하고 있으면 answer++해줌
//그리고 obj의 벨류는 -해줌
//


// lost = [6,4,9]
// 가져온애가 = 5,7 
// sort를 해주면 4는 5를 가져가고 6은 7을 가져감
// 근데 안해주면 6이 5를 체가버림, 4는 원래는 5를 가져갈수 있는데 뺏김, 그래서 최종적으로 6만 얻음

function solution(n, lost, reserve) {
  let answer = n-lost.length // 최소
  let obj = {}
  lost.sort((a,b)=>a-b)
  
  // 1
  for (let i=0; i<reserve.length; i++) {
    obj[reserve[i]] = 1
  }

  // 2
  for (let j=0; j<lost.length; j++) {
    if (obj[lost[j]] === 1) {
      obj[lost[j]] = 0
      answer ++
    }
  }
  
  // 3
  for (let k=0; k<lost.length; k++) {
    if(obj[lost[k]] === 0) continue
    else if (obj[lost[k]-1]) { 
      obj[lost[k]-1] = 0
      answer ++
    }
    else if (obj[lost[k]+1]) { 
      obj[lost[k]+1] = 0
      answer ++
    }
  }
  
  return answer;
}


console.log(solution(5,[2, 4],[1, 3, 5])) // 5
console.log(solution(5,[2, 4],[3])) // 4
console.log(solution(3,[3],[1])) // 2

File: 프로그래머스/1/42862. 체육복/체육복.py
# 1트 - 실패
    # 공통요소를 처음부터 제거하고 시작했는데,
    # 문제의 핵심은 lost의 수를 최대로 줄이는 것
    # 근데, 아래코드처럼 하면
    # print(solution(5,[1,2,3],[3,4,5]))를 대처할 수가 없음..
    # 
# def solution(n, lost, reserve):
#     # 1. lost, reserve에 공통 요소 제거
#     for i in lost:
#         if i in reserve:
#             lost.remove(i)
    
#     # 0. 정렬
#     lost.sort()
#     reserve.sort()
    
#     # 2. reserve순회하면서, lost커버쳐주기 -> lost 최대한 줄이기
#     for j in reserve:
#         if ((j-1) in lost):
#             lost.remove(j-1)
#         elif ((j+1) in lost):
#             lost.remove(j+1)

#     return n-len(lost)


# ---
# 2트 - 실패..(25/30)
    # 뭐가 또 부족한가..
    # 어떤 반례가 있나? - 아 이번엔 또 반대로, 같은건 처리안했네 ㅋㅋ
    # -> 근데 여전히 (27/30)임..
    # 아. 공통요소빼는게 맞네 다시 생각해보니까..
        # 자기가 도난당하면, 다른 사람 못 빌려주잖아..

# def solution(n, lost, reserve):
#     # 0. 정렬
#     lost.sort()
#     reserve.sort()
    
#     # 1. reserve순회하면서, lost커버쳐주기 -> lost 최대한 줄이기
#     for j in reserve:
#         if ((j-1) in lost): # j+1보다 j-1먼저 하는게 중요
#             lost.remove(j-1)
#         elif ((j) in lost): # 이거 추가
#             lost.remove(j)
#         elif ((j+1) in lost):
#             lost.remove(j+1)

#     return n-len(lost)


# ---
# 3트 - 1트의 공통요소 제거 코드 추가 & lost커버쳐줄때, j in lost조건 빼기
def solution(n, lost, reserve):
    # 0. 정렬
    lost.sort()
    reserve.sort()

    
    # 1-0. for문에서 lost.remove(i)하면 for문이 제대로 돌아가겠냐..?
        # 근데, 얕은복사로 해야, 원본도 안 바뀜
    tmp = lost[:] # tmp=lost하면 안됨
    # 1. lost, reserve에 공통 요소 제거    
    for i in lost:
        if i in reserve:
            tmp.remove(i)
            reserve.remove(i) # 이걸 1트에서 안했었음
    
    # 2. reserve순회하면서, lost커버쳐주기 -> tmp 최대한 줄이기
    for j in reserve:
        if ((j-1) in tmp):
            tmp.remove(j-1)
        elif ((j+1) in tmp):
            tmp.remove(j+1)

    return n-len(tmp)


print(solution(3,[3],[1])) # 2
print(solution(5,[1,2,3],[3,4,5])) # 3
print(solution(5,[4,5],[3,4])) # 4
print(solution(5,[1,2,3],[1,2,3])) # 5